apps/pkcs12.c

```diff
@@ -901,7 +901,11 @@ int dump_certs_keys_p12(BIO *out, const PKCS12 *p12, const char *pass,
         } else if (bagnid == NID_pkcs7_encrypted) {
             if (options & INFO) {
                 BIO_printf(bio_err, "PKCS7 Encrypted data: ");
-                alg_print(p7->d.encrypted->enc_data->algorithm);
+                if (p7->d.encrypted == NULL) {
+                    BIO_printf(bio_err, "<no data>\n");
```

Comment:
not sure how to fix this, I've added a check for NULL dereference and the error is then handled in `PKCS12_unpack_p7encdata()`, output then looks like:

```
user@user:~/openssl$ openssl pkcs12 -in test/recipes/80-test_pkcs12_data/bad2.p12 -passin pass: -info
MAC: sha256, Iteration 2048
MAC length: 32, salt length: 8
PKCS7 Encrypted data: <no data>
Error outputting keys and certificates
80CB79DB737F0000:error:11800065:PKCS12 routines:PKCS12_unpack_p7encdata:decode error:crypto/pkcs12/p12_add.c:163:
```

Reply:
Since `PKCS12_unpack_p7encdata` is going to be called anyway, how about call it first and then print the algorithm, like:

~~~
...
bags = PKCS12_unpack_p7encdata(p7, pass, passlen);
if (bags == NULL)
    goto err;
alg_print(p7->d.encrypted->enc_data->algorithm);
....
~~~

Reply:
fixed as suggested, now the output looks like:

```
user@user:~/openssl$ openssl pkcs12 -in test/recipes/80-test_pkcs12_data/bad2.p12 -passin pass: -info
MAC: sha256, Iteration 2048
MAC length: 32, salt length: 8
Error outputting keys and certificates
800B080A977F0000:error:11800065:PKCS12 routines:PKCS12_unpack_p7encdata:decode error:crypto/pkcs12/p12_add.c:163:
```

does this look better now?

Reply:
Well, actually I'd agree with @t8m that the previous version was already okay.

Reply:
ok, reverted the commit to the previous version